### PR TITLE
BE: Fix serde selection for explicitly configured serdes without topic patterns

### LINKED ISSUE #1854: Message key deserialization fails with Confluent context path in Schema Registry
### Issue submitter TODO list

- [x] I've looked up my issue in [FAQ](https://ui.docs.kafbat.io/faq/common-problems)
- [x] I've searched for an already existing issues [here](https://github.com/kafbat/kafka-ui/issues)
- [x] I've tried running `main`-labeled docker image and the issue still persists there
- [x] I'm running a supported version of the application which is listed [here](https://github.com/kafbat/kafka-ui/blob/main/.github/SECURITY.md)

### Describe the bug (actual behavior)

Message key deserialization fails when Schema Registry is configured with a Confluent context path. This issue was originally reported in the provectus/kafka-ui repo and remains unresolved in the latest version of Kafbat UI.

Original issue: https://github.com/provectus/kafka-ui/issues/3792

When using a context-scoped SR URL (e.g. `http://schemaregistry:8081/contexts/.mycontext`), message 

### Expected behavior

Message key should deserialize when opening a topic.

### Your installation details

- We used helm chart version 1.5.0 as well latest one

### Steps to reproduce

1. Open Kafka UI
2. Select a topic with messages on it
3. You will see the message key is not deserialized, and when clicking on the actual message key, it still shows the special chars.

### Screenshots

_No response_

### Logs

_No response_

### Additional context

1. [We tried given solution provided at comment adding separate serde](https://github.com/provectus/kafka-ui/issues/3792#issuecomment-1559669186), We're able to deserilesed value but problem with key still persist, 

Here is **values.yaml** used for kafbat-ui helm chart version **1.5.0** and **latest**
```
yamlApplicationConfig:
  kafka:
    clusters:
      - name: my-cluster
        bootstrapServers: kafka:9092
        # schemaRegistry: http://schemaregistry:8081/contexts/.mycontext
        schemaRegistry: http://schemaregistry:8081
        serde:
          - name: SchemaRegistry (mycontext)
            className: io.kafbat.ui.serdes.builtin.sr.SchemaRegistrySerde
            properties:
              url: http://schemaregistry:8081/contexts/.mycontext
              keySchemaNameTemplate: ":.mycontext:%s-key"
              schemaNameTemplate: ":.mycontext:%s-value"
```

Image

### PR DESCRIPTION
**Pull request by Haarolean**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing application instances)

**What changes did you make?** (Give an overview)
Fixes #1854

### Background: the provectus issue

This issue was originally reported as provectus/kafka-ui#3792. Users discovered that message **key** deserialization broke when using Confluent [schema contexts](https://docs.confluent.io/platform/current/schema-registry/schema-linking-cp.html#using-a-base-context-path) (e.g., `http://schemaregistry:8081/contexts/.mycontext`). Values deserialized fine, but keys showed raw bytes.

Merged PR provectus/kafka-ui#3857 ("Minor serdes fixes") and closed the issue. However, that PR only added **resilience** — wrapping schema lookups in 404 handlers so the app wouldn't crash on missing schemas. It did not fix the actual serde selection problem. The team also suggested a workaround: configure separate custom serdes per context with explicit `keySchemaNameTemplate` / `schemaNameTemplate` properties.

The workaround partially worked (values deserialized), but key deserialization remained broken. The issue was closed prematurely, and the underlying bug carried over into the kafbat fork as #1854.

### Root cause

The serde selection mechanism in `ClusterSerdes.findSerdeByPatternsOrDefault()` iterates over registered serdes and for each one:

1. Checks `couldBePreferable(topic, type)` — whether the serde considers itself suitable
2. Checks the topic pattern (`topicKeyPattern` / `topicValuePattern`) — whether the topic matches

When a user configures a custom serde without setting `topicKeysPattern`/`topicValuesPattern` (the common case for context-scoped SR serdes), these patterns are `null`. The old code treated `null` as "never match":

```java
if (pattern != null && pattern.matcher(topic).matches() && ...) {
    return Optional.of(serdeInstance);
}
```

So even though the custom serde correctly reported `couldBePreferable = true`, it was always skipped. The system fell back to the auto-configured root SR serde (which can't resolve context-scoped schema IDs) or ultimately to `StringSerde` (raw bytes).

### Options considered

1. **Treat null patterns as "match all" universally** — rejected because auto-configured built-in serdes (StringSerde, Int32Serde, etc.) also have null patterns and a default `couldBePreferable()` returning `true`. This would cause StringSerde to win for every topic since it's iterated first.

2. **Require users to always set `topicKeysPattern: .*`** — rejected as poor UX. If a user defines a custom serde with context templates, it should just work without requiring a non-obvious extra config property.

3. **Distinguish explicitly configured vs auto-configured serdes** (chosen) — add an `explicitlyConfigured` flag to `SerdeInstance`. For explicitly configured serdes, null patterns mean "match all topics". For auto-configured built-in serdes, null patterns retain the old "never match" behavior.

### Changes

- **`SerdeInstance`**: Added `explicitlyConfigured` field
- **`ClusterSerdes`**: Updated `findSerdeByPatternsOrDefault()` — null pattern on explicitly configured serdes matches any topic
- **`SerdesInitializer`**: All `SerdeInstance` constructors updated — serdes from user config get `true`, auto-configured/internal ones get `false`
- **`ClusterSerdesTest`** (new): Tests for the new selection behavior

**Is there anything you'd like reviewers to focus on?**

Whether there are edge cases where an explicitly configured serde with null patterns matching all topics could cause unexpected serde selection order issues.

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)

- [ ] No need to
- [ ] Manually (please, describe, if necessary)
- [x] Unit checks
- [ ] Integration checks
- [ ] Covered by existing automation

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [x] My changes generate no new warnings (e.g. Sonar is happy)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

Check out [Contributing](https://github.com/kafbat/kafka-ui/blob/main/.github/CONTRIBUTING.md) and [Code of Conduct](https://github.com/kafbat/kafka-ui/blob/main/.github/CODE-OF-CONDUCT.md)

**A picture of a cute animal (not mandatory but encouraged)**

## Summary by CodeRabbit

* **Improvements**
  * Refined serialization/deserialization format selection to properly prioritize explicitly configured options when selecting formats for topics.
  * Enhanced pattern-based matching for topic-specific format configuration.

* **Tests**
  * Added comprehensive test coverage for serde selection behavior, including explicit configuration and regex pattern matching scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/kafbat/kafka-ui/pull/1856)

### CODE REVIEW

File: api/src/main/java/io/kafbat/ui/emitter/MessageFilters.java [hunk 1]
```diff
@@ -49,14 +51,32 @@ public class MessageFilters {
...
+    return msg -> msgContains(msg, string)
+        || (escapedUpper != null && msgContains(msg, escapedUpper))
+        || (escapedLower != null && msgContains(msg, escapedLower));
+  }
+
+  @Nullable
+  static String escapeNonAscii(String input) {
+    String escaped = StringEscapeUtils.escapeJson(input);
+    return escaped.equals(input) ? null : escaped;
+  }
```
[the remark this answers is not in the document]
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/emitter/MessageFilters.java:74]: @coderabbitai outdated
